applications/DEMApplication/python_scripts/DEM_analysis_stage.py

```diff
@@ -1072,41 +1072,110 @@ def MeasureSphereForGettingPackingProperties(self, radius, center_x, center_y, c
 
-                    if center_to_sphere_distance_0 < (radius - r_0):
+                    if center_to_sphere_distance_0 < (radius - r_0) or center_to_sphere_distance_1 < (radius - r_1):
+                        d = ((x_0 - x_1)**2 + (y_0 - y_1)**2 + (z_0 - z_1)**2)**0.5
+                        if d <= (r_0 + r_1):
+                            vector1 = np.array([x_1 - x_0 , y_1 - y_0, z_1 - z_0])
+                            v1_norm = np.linalg.norm(vector1)
+                            if v1_norm:
+                                vector1_unit = vector1 / v1_norm
+                            tensor = np.outer(vector1_unit, vector1_unit)
+                            total_tensor += tensor
+                            total_contact_number += 1
+                        else:
+                            continue
 
-                        vector1 = np.array([x_1 - x_0 , y_1 - y_0, z_1 - z_0])
-                        v1_norm = np.linalg.norm(vector1)
-                        if v1_norm:
-                            vector1_unit = vector1 / v1_norm
-                        tensor = np.outer(vector1_unit, vector1_unit)
-                        total_tensor += tensor
-                        total_contact_number += 1
-
-                    elif center_to_sphere_distance_1 < (radius - r_1):
-
-                        vector1 = np.array([x_1 - x_0 , y_1 - y_0, z_1 - z_0])
-                        v1_norm = np.linalg.norm(vector1)
-                        if v1_norm:
-                            vector1_unit = vector1 / v1_norm
-                        tensor = np.outer(vector1_unit, vector1_unit)
-                        total_tensor += tensor
-                        total_contact_number += 1
-                
                 if total_contact_number:
                     measured_fabric_tensor = total_tensor / total_contact_number
+
                 else:
                     measured_fabric_tensor = np.empty((3, 3))
 
-                deviatoric_tensor = 4 * (measured_fabric_tensor - 1/3 * np.eye(3)) 
+                deviatoric_tensor = 4 * (measured_fabric_tensor - 1/3 * np.eye(3))
 
                 second_invariant_of_deviatoric_tensor = (0.5 * np.sum(deviatoric_tensor * deviatoric_tensor))**0.5
 
+                fabric_tensor_trace = (measured_fabric_tensor[0][0] + measured_fabric_tensor[1][1] + measured_fabric_tensor[2][2])/3
+
                 eigenvalues, eigenvectors = np.linalg.eig(measured_fabric_tensor)
-                
-                return eigenvalues, second_invariant_of_deviatoric_tensor, measured_fabric_tensor
+
+                return eigenvalues, second_invariant_of_deviatoric_tensor, [measured_fabric_tensor[0][0],measured_fabric_tensor[1][1],measured_fabric_tensor[2][2]], fabric_tensor_trace
 
             else:
                 raise Exception('The \"ContactMeshOption\" in the [ProjectParametersDEM.json] should be [True].')
+        if type == "conductivity_tensor":
```

Comment:
hmmm it is expected to have in the master this large pieces of code in the base analysis stage of the DEM? Honestly (as a DEM user), this node loops are extremely slow in python and IMO they should be implemented in a `process` (either python or preferably c++).

Why do we have these expensive operations in the AnalysisStage?

Reply:
Do you know why is this here @ChengshunShang1996

Reply:
I agree with your opinion. It is better to put this large piece of code in DEM_procedures.py instead of DEM_analysis_stage. This piece of code is designed for measurement and post-processing. So it will/should be called every output step (1e3-1e4 steps). Thus it should not slow down the DEM simulation a lot.

Reply:
We can do a cleanup and move this to DEM_procedures.py or C++ in the near future. Designing it in Python at the beginning was mainly for convenience and it will not slow down the simulations in most of the cases. It grows up to a large piece of code gradually, thus it may be better to move it to a better place.

Reply:
I am currently working on a branch where I have my 2D implementations of packing evaluation (fabric, stress, conductivity tensors, etc), which is a utility that is used for RVE analysis. I plan to make a PR of this in a near future. Maybe it could be used to accommodate such implementations.

Reply:
Those functions are similar to those you mentioned. It would be great if we could put them together in a Python Class or similar concept.

Reply:
They are c++ [utilities](https://github.com/KratosMultiphysics/Kratos/blob/hierarchical-multiscale/applications/DEMApplication/custom_utilities/rve_utilities.h) in this [branch](https://github.com/KratosMultiphysics/Kratos/tree/hierarchical-multiscale).

Reply:
Cool! I had a quick glance at it. It looks like you measure those properties and write them into files, right? In Python, multiple parameters can be returned by one measurement function, and those parameters can be used to control the simulation.

Reply:
It is similar to the fabric tensor implementation, if that is included in DEM_analysis_stage.py, why not this one?
Anyway, you can set the calculation to False by default to avoid using it

Reply:
I just followed the implementation of the fabric tensor that ChengShun did.